### PR TITLE
Changes to kll

### PR DESCRIPTION
**Pull request by leerho**

### CODE REVIEW

**Comment by leerho** [comment]
This, unfortunately, is a massive PR, and will be difficult to review.  Nevertheless, the main goals of this PR:
- Fix the getQuantile(), getRank(), getPMF(), getCDF() for the REQ and KLL sketches so that they properly handle both INCLUSIVE and NON_INCLUSIVE queries.  ( I haven't done this for classic quantiles, but this is already so big I decided to get this PR done and do another PR for classic.
- This also involved fixing some bugs in the InequalitySearch and GenericInequalitySearch classes.  These bugs only involved sketches with only a single item.
- This also involved streamlining the Sorted View classes. By leveraging the SV classes for getRank, getCDF, and getPMF, I was able to eliminate over 300 lines of code.  I haven't done this for REQ, as the savings was not quite as large, but may still decide to do this in REQ before we finally push to master. 
- Change all references to "item" to "value" for numeric valued quantile sketches. We decided to use "item" only for the generic sketches.  This was huge as there were hundreds of uses of "item" in almost every class.
- i did extensive rewrite of javadocs and code comments where I found them confusing or incomplete.

**Comment by leerho** [comment]
This is now FROZEN and ready for review.  No more changes!

File: src/main/java/org/apache/datasketches/QuantileSearchCriteria.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+   *
+   * <p><b>Definition of NON_INCLUSIVE <i>getRank(q)</i> search:</b><br>
+   * Given quantile <i>q</i>, return the rank, <i>r</i>, of the largest quantile that is
+   * strictly less than <i>q</i>.</p>
+   *
+   * <p>If there is no quantile value that is strictly less than <i>q</i>,
+   * the method will return a rank of zero.</p>
+   *
+   */
+  NON_INCLUSIVE,
```
Comment (davecromberge) [inline src/main/java/org/apache/datasketches/QuantileSearchCriteria.java:61]: Would `exclusive` be preferred, or is non-inclusive used for duality?

File: src/main/java/org/apache/datasketches/req/ReqSketch.java [hunk 2]
```diff
@@ -336,23 +323,17 @@ public double getRankLowerBound(final double rank, final int numStdDev) {
...
   public double[] getRanks(final float[] values) {
-    return getRanks(values, ltEq);
+    return getRanks(values, ltEq == true ? INCLUSIVE : NON_INCLUSIVE);
   }
 
   @Override
-  public double[] getRanks(final float[] values, final boolean inclusive) {
+  public double[] getRanks(final float[] values, final QuantileSearchCriteria searchCrit) {
     if (isEmpty()) { return null; }
+    refreshSortedView();
```
Comment (davecromberge) [inline src/main/java/org/apache/datasketches/req/ReqSketch.java:332]: Why is the sorted view refreshed on read-only functions?  Would it be better to offset the cost on updates to the sketch or is this necessary because the sketch may be backed by direct memory and modified behind the scenes?

**Review by davecromberge — APPROVED** [review]
Apologies for the late review.  I think the documentation is clear and I understand the idea behind the changes.  Namely, a unified and consistent behaviour for sorting and searching between the different quantiles sketch implementations.
I spotted a few minor typos along the way, and I have added a commit to correct them.

File: src/main/java/org/apache/datasketches/DoublesSortedView.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+  /**
+   * Returns an array of values where each value is a number in the range [0.0, 1.0].
+   * The size of this array is one larger than the size of the input splitPoints array.
+   *
+   * <p>If <i>isCdf</i> is true, the points in the returned array are monotonically increasing and end with the
+   * value 1.0. Each value represents a point along the cumulative distribution function that approximates
+   * the CDF of the input data stream. Therefore, each point represents the fractional density of the distribution
+   * between from zero. For example, if one of the returned values is 0.5, then the splitPoint corresponding to that
+   * value would be the median of the distribution.</p>
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/main/java/org/apache/datasketches/DoublesSortedView.java:54]: Agreed.  If I change it in this PR it will create a conflict with  the next PR. So I will change it the next PR.

File: src/test/java/org/apache/datasketches/ReflectUtility.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  @Test //Example
+  public void checkCtr() throws Exception {
+    float[] farr = { 10, 20, 30 };
+    long[] larr = { 1, 2, 3 };
+    long n = 3;
+    ReqSketchSortedView reqSV =
+        (ReqSketchSortedView) REQ_SV_CTOR.newInstance(farr, larr, n);
+    float q = reqSV.getQuantile(1.0, INCLUSIVE);
+    assertEquals(q, 30f);
+  }
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/test/java/org/apache/datasketches/ReflectUtility.java:64]: This Reflect Utility is used by the CrossCheckQuantilesTest which lives in the o.a.datasketches package and runs a common suite of tests against all the different quantiles-type sketches, which live in different packages, and some of these tests leverage the package-private constructor that is common for all of the Sorted View implementations.   Thus, the need for reflection.

File: src/test/java/org/apache/datasketches/GenericInequalitySearchTest.java [hunk 5]
```diff
@@ -109,7 +109,7 @@ private void checkBinarySearchFloatLimits(final Float[] arr, final int low, fina
     v = highV + 1;
     res = find(arr, low, high, v, LT, comparator);
     println(desc(arr, low, high, v, res, LT, comparator));
-    assertEquals(res, high);
+    assertEquals(res, high); //??
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/test/java/org/apache/datasketches/GenericInequalitySearchTest.java:112]: Thank you!  Will fix with the next PR.

File: src/test/java/org/apache/datasketches/CrossCheckQuantilesTest.java [hunk 6]
```diff
@@ -19,300 +19,219 @@
...
 
 public class CrossCheckQuantilesTest {
 
-  enum SkType { REQ, REQ_SV, REQ_NO_DEDUP, KLL, CLASSIC }
+  final int k = 32; //all sketches are in exact mode
 
-  enum PrimType { DOUBLE, FLOAT }
+  //These test sets are specifically designed to test some tough corner cases so don't mess with them
+  //  unless you know what you are doing.
+  //These sets must start with 10 and be multiples of 10.
```
[the remark this answers is not in the document]
Comment (leerho) [reply src/test/java/org/apache/datasketches/CrossCheckQuantilesTest.java:45]: Good point.  Will do in next PR.

File: src/main/java/org/apache/datasketches/QuantileSearchCriteria.java (hunk 1 above)
Comment (leerho) [reply src/main/java/org/apache/datasketches/QuantileSearchCriteria.java:61]: Now why didn't I think of that!  Will fix in next PR.

File: src/main/java/org/apache/datasketches/req/ReqSketch.java (hunk 2 above)
Comment (leerho) [reply src/main/java/org/apache/datasketches/req/ReqSketch.java:332]: Building the sorted view can be an expensive operation.  Once it has been built, and if the sketch has not been updated or been the target of a merge, all subsequent calls to the query methods (of which, getRanks is one) are quite fast.  However, if the user does a getRanks(); update(); getRanks(), the update() call invalidates the internal state of the sketch, so the sorted view needs to be rebuilt.

Thus each of the query calls must check if the internal state of the sketch has changed.  If not, this refreshSortedView() does nothing.
